### PR TITLE
learing with svm + expectation lemmas

### PR DESCRIPTION
**Pull request by RaliDardjonova**
Theories about soft-svm, strong convexity, convex learning and some lemmas about their expectations. I will add comments in the future.

### CODE REVIEW

File: RLM_learn.thy [hunk 1]
```diff
@@ -0,0 +1,189 @@
+theory RLM_learn
```
Comment (maxhaslbeck) [inline RLM_learn.thy:3]: please adhere to [theory documentation guidelines](https://isabelle.systems/conventions/theory_documentation.html), at least add creator, summary and main definitions/theorems 

maybe spell out RLM, and use upper Snake Case

File: RLM_learn.thy [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+  imports  "LearningTheory" "Strong_Convexity"
+begin
+
+
+
+definition pred_err_loss :: "('a \<times> 'b) pmf \<Rightarrow>  ('c  \<Rightarrow> ('a \<times> 'b) \<Rightarrow>  real)
```
Comment (maxhaslbeck) [inline RLM_learn.thy:38]: for definitions please add a description and possibly a reference, you might add a .bib file in the ROOT file, and reference it using the antiquotation @{cite XXX}

File: Strong_Convexity.thy [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+end
+
+instantiation "fun" :: (ab_semigroup_add, ab_semigroup_add) ab_semigroup_add
+begin
+
+instance proof
+  fix x y z :: "'a => 'b"
+  show "x + y + z = x + (y + z)"
+    unfolding fun_plus_def 
+    by (simp add: linordered_field_class.sign_simps(1))
```
Comment (maxhaslbeck) [inline Strong_Convexity.thy]: in Isabelle2020 this fact does not exist, are you using Isabelle2019? please migrate!

File: Strong_Convexity.thy [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+instance proof
+  fix x y z :: "'a => 'b"
+  show "x + y + z = x + (y + z)"
+    unfolding fun_plus_def 
+    by (simp add: linordered_field_class.sign_simps(1)) 
+next
+  fix x y :: "'a => 'b"
+  show "x + y = y + x"
+    unfolding fun_plus_def
+    by (simp add: linordered_field_class.sign_simps(2))
```
Comment (maxhaslbeck) [inline Strong_Convexity.thy]: in Isabelle2020 this fact does not exist, are you using Isabelle2019? please migrate!

File: Strong_Convexity.thy [hunk 5]
```diff
@@ -0,0 +1,389 @@
+theory Strong_Convexity
+  imports Main "HOL-Analysis.Analysis" "HOL-Analysis.Convex" 
+begin
+
+definition strong_convex_on :: "'a::euclidean_space set\<Rightarrow> ('a \<Rightarrow> real) \<Rightarrow> real \<Rightarrow> bool"
+  where "strong_convex_on s f k \<longleftrightarrow>
```
Comment (maxhaslbeck) [inline Strong_Convexity.thy]: for sets you might want to use capital letters, i.e. S

File: Strong_Convexity.thy [hunk 6]
```diff
@@ -0,0 +1,389 @@
+theory Strong_Convexity
```
Comment (maxhaslbeck) [inline Strong_Convexity.thy:3]: please adhere to [theory documentation guidelines](https://isabelle.systems/conventions/theory_documentation.html), at least add creator, summary and main definitions/theorems

File: Strong_Convexity.thy [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+  for x y :: "'a::euclidean_space"
+  by (smt inner_commute inner_left_distrib power2_norm_eq_inner scaleR_2)
+
+lemma help2_31 : "norm (x - y)^2 = norm x ^ 2 - 2 *\<^sub>R (inner x y) + norm y ^ 2" 
+  for x y :: "'a::euclidean_space"
+  using help2_3
+  by (simp add: inner_commute inner_diff_right power2_norm_eq_inner)
+
+
+lemma help2_2 : "(norm (u  *\<^sub>R x +  v  *\<^sub>R y))^2 = norm (u *\<^sub>R x)^2 + (2 * u * v) *\<^sub>R (inner x y) + norm (v *\<^sub>R y)^2 "
```
Comment (maxhaslbeck) [inline Strong_Convexity.thy]: try to adhere to "Lines must not be longer than 100 symbols (as indicated by the blue line in jEdit)."

File: Strong_Convexity.thy [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+
+lemma help2_4: "norm (u *\<^sub>R x)^2  = u^2 * norm(x)^2"
+proof -
+  have "abs(u)^2 = u^2" by simp
+  then show "norm (u *\<^sub>R x)^2  = u^2 * norm(x)^2" 
+    using norm_scaleR power2_eq_square
+    by (simp add: power_mult_distrib)
+qed
+
+lemma sq_norm_strong_convex: "strong_convex_on s (\<lambda> w. k * norm(w) * norm(w)) (2*k)"
```
Comment (maxhaslbeck) [inline Strong_Convexity.thy]: you might want to add an introduction rule for a new concept, here: strong_convex_onI:
```
lemma strong_convex_onI:
  assumes 
    "⋀x y u v. x∈s ⟹ y∈s ⟹  u≥0 ⟹  v≥0 ⟹ u + v = 1 ⟹
     f (u *⇩R x + v *⇩R y) ≤ u * f x + v * f y - (k/2) * u * v * norm(x-y) * norm(x-y)"
  shows "strong_convex_on s f k"
  using assms
  unfolding strong_convex_on_def  
  by blast 
```
and then use it
similarly you might want to have a destruction rule `strong_convex_onD`

File: Strong_Convexity.thy [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+  for s :: "'a::euclidean_space set"
+proof -
```
Comment (maxhaslbeck) [inline Strong_Convexity.thy]:
```suggestion
proof (rule strong_convex_onI)
```
then you can save restating the lemma in line 37 and the blast call in the end,

File: Strong_Convexity.thy [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+  imports Main "HOL-Analysis.Analysis" "HOL-Analysis.Convex" 
+begin
+
+definition strong_convex_on :: "'a::euclidean_space set\<Rightarrow> ('a \<Rightarrow> real) \<Rightarrow> real \<Rightarrow> bool"
+  where "strong_convex_on s f k \<longleftrightarrow>
+    (\<forall>x\<in>s. \<forall>y\<in>s. \<forall>u\<ge>0. \<forall>v\<ge>0. u + v = 1 \<longrightarrow>
+     f (u *\<^sub>R x + v *\<^sub>R y) \<le> u * f x + v * f y - (k/2) * u * v * norm(x-y) * norm(x-y) )" 
+
+
+lemma help2_3 : "norm (x+y)^2 = norm x ^ 2 + 2 *\<^sub>R (inner x y) + norm y ^ 2"
```
Comment (maxhaslbeck) [inline Strong_Convexity.thy]: you might use the syntax for inner: `x ∙ y`, e.g look at the lemma statement in the output/proof state panel to see the correct syntax. this might make your theories shorter and more readable

File: Strong_Convexity.thy [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+      then have "?L \<ge> 0" using metric_LIM_le_zero using 7 by blast
+      then show ?thesis by auto
+    qed
+  qed
+qed
+
+lemma strong_conv_if_eq: " f = g \<Longrightarrow> strong_convex_on s f k \<Longrightarrow> strong_convex_on s g k"
+  using  HOL.subst by auto
+
+lemma strong_conv_then_conv:
```
Comment (maxhaslbeck) [inline Strong_Convexity.thy]: I wondered why this lemma is down here. you could move that lemma upwards, directly after the definition of strong_convex_on and its intro/dest rules

File: Strong_Convexity.thy [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+
+lemma strong_convex_sum: "strong_convex_on s f k \<and> convex_on s g  \<longrightarrow> 
+                            strong_convex_on s ( f + g) k"
+proof 
+  assume "strong_convex_on s f k \<and> convex_on s g"
+  then show "strong_convex_on s (f + g) k"
+  proof
+    have "strong_convex_on s f k" using `strong_convex_on s f k \<and> convex_on s g` by simp
+    have "convex_on s g" using `strong_convex_on s f k \<and> convex_on s g` by simp
+    have  "(\<forall>x\<in>s. \<forall>y\<in>s. \<forall>u\<ge>0. \<forall>v\<ge>0. u + v = 1 \<longrightarrow>
```
Comment (maxhaslbeck) [inline Strong_Convexity.thy]: here you can use the introduction rule again

File: Strong_Convexity.thy [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+lemma convex_fun_add:
+  assumes "convex_on s f" "convex_on s g"
+  shows "convex_on s (f + g)"
+proof - 
+  have "(f + g) = (\<lambda> x. f x + g x)" using fun_plus_def by auto
+  moreover have "convex_on s (\<lambda>x. f x + g x)" using assms convex_on_add by auto
+  ultimately show "convex_on s (f + g)" by auto
+qed
+
+lemma strong_convex_sum: "strong_convex_on s f k \<and> convex_on s g  \<longrightarrow>
```
Comment (maxhaslbeck) [inline Strong_Convexity.thy]: prefer to use the long lemma format
```suggestion
lemma strong_convex_sum:
  assumes scf: "strong_convex_on s f k" and cg: "convex_on s g"
  shows  "
```
then you can comfortably use the facts in the proof

follow the [style guide](https://isabelle.systems/conventions/style.html)

File: Strong_Convexity.thy [hunk 14]
```diff
@@ -0,0 +1,389 @@
...
+    proof -
+      have "(\<forall>x\<in>s. \<forall>y\<in>s. \<forall>u\<ge>0. \<forall>v\<ge>0. u + v = 1 \<longrightarrow>
+     f (u *\<^sub>R x + v *\<^sub>R y) \<le> u * f x + v * f y - (k/2) * u * v * norm(x-y)^2)"
+        using assms(1) unfolding  strong_convex_on_def 
+        by (simp add: power2_eq_square mult.commute mult.left_commute)
+      then have 0:" \<forall>u\<ge>0. \<forall>v\<ge>0. u + v = 1 \<longrightarrow>
+     f (u *\<^sub>R w + v *\<^sub>R x) \<le> u * f w + v * f x - (k/2) * u * v * norm(w-x)^2"
+        using assms(2) assms(4) by blast
+
+      have  "\<forall>u>0. \<forall>v\<ge>0. u + v = 1 \<longrightarrow>
```
Comment (maxhaslbeck) [inline Strong_Convexity.thy:336]: instead of writing down stuff twice you can also do the following
```
{
 fix a
 assume A
 have B
 proof -
  ...
 qed
} note somename= this 
```

File: RLM_learn.thy [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+ \<Rightarrow> 'c \<Rightarrow> real" where
+  "pred_err_loss D l h = measure_pmf.expectation D (\<lambda> z. (l h z))"
+
+definition train_err_loss :: " (nat \<Rightarrow> ('a * 'b)) \<Rightarrow> nat \<Rightarrow>
+('c \<Rightarrow> ('a \<times> 'b) \<Rightarrow> real) \<Rightarrow> 'c \<Rightarrow>  real" where
+  "train_err_loss S n l h = sum (\<lambda>i. l h (S i) ) {0..<n} / n"
+
+lemma train_err_loss_nn: "(\<forall>i\<in>{0..<n}. l h (S i) \<ge>0) \<Longrightarrow> train_err_loss S n l h \<ge> 0"
```
Comment (maxhaslbeck) [inline RLM_learn.thy]: long lemma format spares writing the assumptions twice (see [style guide](https://isabelle.systems/conventions/style.html) )

File: RLM_learn.thy [hunk 16]
```diff
@@ -0,0 +1,189 @@
+theory RLM_learn
+  imports  "LearningTheory" "Strong_Convexity"
+begin
+
+
```
Comment (maxhaslbeck) [inline RLM_learn.thy:7]: section and document your theory.
use command `section`, and `subsection` to structure your theory! Readers then can use the Sidekick panel in Isabelle/JEdit to navigate the theory more comfortably!

File: RLM_learn.thy [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+qed
+
+
+
+text\<open>Show L_s(w) is convex over H, when the loss function is convex over H\<close>
+lemma train_err_loss_if_convex: "(\<forall>i \<in>{0..<n} . convex_on H (\<lambda> h. l h (S i))) \<Longrightarrow>
+   convex_on H (train_err_loss S n l)"
+proof -
+  assume 1: "(\<forall>i \<in> {0..<n}. convex_on H (\<lambda> h. l h (S i)))" 
+  then have 2: "convex_on H (\<Sum>i\<in>{0..<n}. (\<lambda> h. l h (S i)))"
```
Comment (maxhaslbeck) [inline RLM_learn.thy]: can you generalize lemma ``2`` and move it out?

File: RLM_learn.thy [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+  "ridge_mine S' k' = (SOME h. h \<in> ridge_argmin S' k')"
+
+definition swapped_S :: "(nat \<Rightarrow> ('b * 'c)) \<Rightarrow> nat \<Rightarrow> nat  \<Rightarrow> (nat \<Rightarrow> ('b * 'c))" where
+  "swapped_S S' i m =  S'(i:= S' m, m:= S' i) "  
+
+definition truncated_S :: "(nat \<Rightarrow> ('b * 'c))  \<Rightarrow> nat \<Rightarrow>(nat \<Rightarrow> ('b * 'c))" where
+  "truncated_S S' m =  S'(m := undefined) "
+
+text\<open>S_index is a set where the i-th data point in S is replaced with an arbitrary one\<close>
+definition S_index :: "(nat \<Rightarrow> ('b \<times> 'c)) \<Rightarrow> nat \<Rightarrow> ('b \<times> 'c) \<Rightarrow> (nat \<Rightarrow> ('b * 'c))" where
```
Comment (maxhaslbeck) [inline RLM_learn.thy:187]: why do you define ``S_index`` ? it is the same as ``fun_upd``

File: RLM_learn.thy [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+
+lemma sample_in_D: "\<forall>S \<in> (Samples m D). \<forall>i\<in>{0..<m}. S i \<in>  D" 
+   unfolding Samples_def  using  set_pmf_Pi_pmf 
+   by (metis finite_atLeastLessThan)
+
+
+lemma train_err_loss_convex: "\<forall>S \<in> (Samples n D).convex_on H (train_err_loss S n l)"
+  using train_err_loss_if_convex convll sample_in_D  by blast
+
+lemma ridge_strong_convex: "\<forall>S \<in> (Samples n D). strong_convex_on H (ridge_fun S k) (2*k)"
```
Comment (maxhaslbeck) [inline RLM_learn.thy]: prefer to use pure all quantifier over HOL.quantifier, also use long lemma format!

File: Strong_Convexity.thy (hunk 11 above)
Comment (maxhaslbeck) [inline Strong_Convexity.thy]: for lemma names you should spell out constant names, `conv` often stands for an equation (e.g. `fst_conv`). also use the convention to write `conclusion_if_premise`.
`convex_on_if_strong_convex_on` is a suggestion

File: RLM_learn.thy [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+  using strong_conv_then_conv ridge_strong_convex k_pos by smt
+
+
+lemma convex_has_min: 
+  fixes f :: " _ \<Rightarrow> real"
+  assumes "convex_on H f"
+  shows "\<exists>x\<in>H. is_arg_min f (\<lambda>z. z\<in>H) x" 
+  sorry
+
+lemma in_argmin:
```
Comment (maxhaslbeck) [inline RLM_learn.thy]: too general name

File: rlm_13_02_lemma.thy [hunk 21]
```diff
@@ -0,0 +1,965 @@
+theory rlm_13_02_lemma
```
Comment (maxhaslbeck) [inline rlm_13_02_lemma.thy]: please adhere to [theory documentation guidelines](https://isabelle.systems/conventions/theory_documentation.html), at least add creator, summary and main definitions/theorems 

Maybe spell out RLM, and use upper Snake Case
Which lemma is is lemma 13.2 in this theory?

Section and document the theory!

File: rlm_13_02_lemma.thy [hunk 22]
```diff
@@ -0,0 +1,965 @@
...
+  fixes f ::"(  _ \<Rightarrow> real)"
+  fixes g ::"(  _ \<Rightarrow> real)"
+  assumes "\<forall> x \<in> set_pmf M. f x = g x"
+  shows "integral\<^sup>N M f = integral\<^sup>N M g"
+  by (metis (mono_tags, lifting) assms ennreal_0 mult_not_zero nn_integral_cong 
+      nn_integral_measure_pmf pmf_eq_0_set_pmf)
+
+lemma same_nn_integral_restricted1:
+  fixes f ::"(  _ \<Rightarrow> ennreal)"
+  fixes g ::"(  _ \<Rightarrow> ennreal)"
```
Comment (maxhaslbeck) [inline rlm_13_02_lemma.thy]:
```suggestion
  fixes f g ::"(  _ \<Rightarrow> ennreal)"
```

File: RLM_stable.thy [hunk 23]
```diff
@@ -0,0 +1,1091 @@
+theory RLM_stable
```
Comment (maxhaslbeck) [inline RLM_stable.thy:3]: please adhere to [theory documentation guidelines](https://isabelle.systems/conventions/theory_documentation.html), at least add creator, summary and main definitions/theorems

File: RLM_stable.thy [hunk 24]
```diff
@@ -0,0 +1,1091 @@
...
+      ridge_fun (S_index S i z) k v - ridge_fun (S_index S i z) k u
+    + (l v (S i) - l u (S i))/n  + (l u z - l v z)/n " 
+  proof -
+    have "ridge_fun S k v = (train_err_loss S n l + (\<lambda> w. k * (norm w)^2)) v" by simp
+    moreover have "ridge_fun S k u = (train_err_loss S n l + (\<lambda> w. k * (norm w)^2)) u" by simp
+
+    ultimately  have "ridge_fun S k v - ridge_fun S k u = 
+     (train_err_loss S n l + (\<lambda> w. k * (norm w)^2)) v - 
+      (train_err_loss S n l + (\<lambda> w. k * (norm w)^2)) u" by auto
+    also have "(train_err_loss S n l + (\<lambda> w. k * (norm w)^2)) v -
```
Comment (maxhaslbeck) [inline RLM_stable.thy]: use `... =` instead of restating the left hand side in calculation proofs

File: RLM_stable.thy [hunk 25]
```diff
@@ -0,0 +1,1091 @@
...
+
+    then have "norm (?v - ?u) \<le>  2 * \<rho>  / (n*k)" using \<open>0 < norm (?v - ?u)\<close> by auto
+    then have "\<rho> * norm (?v - ?u) \<le> \<rho> * 2 * \<rho>  / (n*k)" using `\<rho> \<ge> 0`
+      by (metis mult_left_mono semiring_normalization_rules(18) times_divide_eq_right)
+    then have "\<rho> * norm (?v - ?u) \<le>  2 * \<rho> * \<rho>  / (n*k)"
+      by (simp add: semiring_normalization_rules(7))
+    then have "\<rho> * norm (?v - ?u) \<le>  2 * \<rho> ^2  / (n*k)" using power2_eq_square
+      by (metis mult.commute semiring_normalization_rules(19))
+    then show " (l ?v (S i)) - (l ?u (S i)) \<le> (2*\<rho>^2)/(k*n)" using 1 
+      by (simp add: dist_norm mult.commute)
```
Comment (maxhaslbeck) [inline RLM_stable.thy:249]: can you turn this in a calculation proof? it might be more readable that way

File: RLM_stable.thy [hunk 26]
```diff
@@ -0,0 +1,1091 @@
...
+  qed
+qed
+
+end
+
+lemma integral_bellow_const:
```
Comment (maxhaslbeck) [inline RLM_stable.thy]: typo in lemma name.
+ pull such general lemmas into a separate theory, e.g. `VerML_Integral_Misc`

File: RLM_stable.thy [hunk 27]
```diff
@@ -0,0 +1,1091 @@
...
+
+lemma univ_to_exist: "A\<noteq>{} \<Longrightarrow> \<forall>x\<in>A. P x \<Longrightarrow> \<exists>x\<in>A. P x" 
+  by blast
+
+context lipschitz_ridge_and_convex_loss
+begin
+
+
+
+lemma integrable_D:"\<forall> S \<in> (Samples n D). integrable D (\<lambda> z. l (ridge_mine S k) z)"
```
Comment (maxhaslbeck) [inline RLM_stable.thy]: `XXX_D` often stands for destruction rules, choose another, more descriptive name! The integrability does not stem from the D but rather from `l`

File: RLM_stable.thy [hunk 28]
```diff
@@ -0,0 +1,1091 @@
...
+   train_err_loss S n l h + k * norm ( h )^2"
+      using 1 unfolding ridge_fun.simps by auto
+    then show " train_err_loss S n l ?w   \<le>
+   train_err_loss S n l h + k * norm ( h )^2" using k_pos
+      by (smt mult_nonneg_nonneg zero_le_power2)
+  qed
+  then show "train_err_loss S n l (ridge_mine S k) \<le>
+                                train_err_loss S n l h +  k * norm ( h )^2"
+    by blast
+qed
```
Comment (maxhaslbeck) [inline RLM_stable.thy]: fix indentation. you can select some lines in Isabelle/JEdit and then use CTRL+i to let Isabelle auto-indent the section (proof or lemma statement). Often this quickly leads to a rather good indentation result.

File: RLM_stable.thy [hunk 29]
```diff
@@ -0,0 +1,1091 @@
...
+  then have "integral\<^sup>N ?M ?v < \<infinity>" using  integrable_iff_bounded by metis
+  then have 6:"integral\<^sup>N ?M ?u < \<infinity>" using 5 by auto
+
+  have "(\<lambda> S. (train_err_loss S n l (ridge_mine S k))) \<in> borel_measurable ?M" by auto
+   have "integrable ?M  (\<lambda> S. (train_err_loss S n l (ridge_mine S k)))" 
+    using 6 integrable_iff_bounded[of ?M "(\<lambda> S. (train_err_loss S n l (ridge_mine S k)))"]
+    by simp
+  then show "integrable (Samples n D) (\<lambda> S. train_err_loss S n l (ridge_mine S k))" 
+    by auto
+qed
```
Comment (maxhaslbeck) [inline RLM_stable.thy]:
```suggestion
lemma train_err_integrable:
  shows " integrable (Samples n D) (λ S. train_err_loss S n l (ridge_mine S k))"
proof -
  from nnH obtain h where "h∈H" by blast 

  let ?f = " (λ i S. l (ridge_mine S k) (S i))"
  let ?g = "(λ S  z. l (ridge_mine S k) z)"
  let ?M = "(Samples n D)"

  have nn: "⋀S. S∈set_pmf ?M ⟹ ∀i∈{0..<n}. 0 ≤ ?f i S "
    using l_pos sample_in_D min_in_H by blast

  {
    fix S
    assume S: "S∈Samples n D"

    have *: "train_err_loss S n l (ridge_mine S k)
                 ≤ train_err_loss S n l h +  k * norm ( h )^2" 
    proof -
      let ?w = " (ridge_mine S k)"
      have " train_err_loss S n l ?w ≤ train_err_loss S n l h + k * norm ( h )^2"
      proof -
        have "is_arg_min (ridge_fun S k) (λs. s∈H) (ridge_mine S k)"
          unfolding ridge_mine_def unfolding ridge_argmin_def 
          by (metis (full_types) S in_argmin k_pos mem_Collect_eq ridge_argmin_def tfl_some)
        then have "(ridge_fun S k)  (ridge_mine S k) ≤ (ridge_fun S k) h" 
          using `h∈H` 
          by (simp add: is_arg_min_linorder)
        then have "train_err_loss S n l ?w  + k* norm((ridge_mine S k))^2
                   ≤ train_err_loss S n l h + k * norm ( h )^2"
          unfolding ridge_fun.simps by auto
        then show "train_err_loss S n l ?w 
                  ≤ train_err_loss S n l h + k * norm ( h )^2" using k_pos
          by (smt mult_nonneg_nonneg zero_le_power2)
      qed
      then show "train_err_loss S n l (ridge_mine S k) ≤
                                train_err_loss S n l h +  k * norm ( h )^2"
        by blast
    qed

    have " train_err_loss S n l (ridge_mine S k) ≥ 0" 
      by (simp add: train_err_loss_nn nn S)
    then have  "(train_err_loss S n l (ridge_mine S k)) =
          (norm (train_err_loss S n l (ridge_mine S k)))" 
      by simp
    then have 12:" (norm (train_err_loss S n l (ridge_mine S k)))
      ≤ norm (train_err_loss S n l h + k * (norm h)⇧2)" 
      using * by auto
  } note bounded = this

  then have integrable_ridge_fun:"integrable ?M (λ S.  train_err_loss S n l h + k * norm ( h )^2)" 
    apply(intro Bochner_Integration.integrable_add) 
    subgoal by(rule train_err_integrable1[OF `h∈H`])
    subgoal by auto
    done

  show "integrable (Samples n D) (λ S. train_err_loss S n l (ridge_mine S k))" 
    apply(rule Bochner_Integration.integrable_bound)
      apply(rule integrable_ridge_fun)
    using bounded
    by (auto intro: AE_pmfI )   
qed
```

File: RLM_stable.thy [hunk 30]
```diff
@@ -0,0 +1,1091 @@
...
+                     (\<lambda> i. (l (ridge_mine (swapped_S S i n) k) (S i))))"
+    using 5 by auto
+  show ?thesis
+    using train_val_diff replace_one_stable2 integrable_D integrable_Si integrable_swapped 
+      pred_err_integrable  
+        train_err_integrable  by auto
+qed
+
+
+lemma oracle_inequality:
```
Comment (maxhaslbeck) [inline RLM_stable.thy:701]: this is the culmination of this theory, right? please document it and explain what it means, or where one can find the lemma in the informal text book

File: Subgradients.thy [hunk 31]
```diff
@@ -0,0 +1,371 @@
...
+text\<open>Don't need to prove it, if we show it is true for the hinge loss\<close>
+lemma convex_has_subgradient:
+  fixes f :: "'a::euclidean_space \<Rightarrow> real"
+  fixes A :: "'a::euclidean_space set"
+  fixes \<rho> :: "real"
+  assumes "\<rho> \<ge> 0"
+  assumes " open A"
+  assumes "convex A"
+  shows "convex_on A f \<longleftrightarrow> (\<forall> w \<in> A. \<exists> v \<in> A. \<forall> u \<in> A. f (u) \<ge> f(w) + (inner (u-w) v)) "
+  sorry
```
Comment (maxhaslbeck) [inline Subgradients.thy:25]: if the lemma is not used, dont use sorry but oops

File: Subgradients.thy [hunk 32]
```diff
@@ -0,0 +1,371 @@
+theory Subgradients
```
Comment (maxhaslbeck) [inline Subgradients.thy:1]: please adhere to [theory documentation guidelines](https://isabelle.systems/conventions/theory_documentation.html), at least add creator, summary and main definitions/theorems

File: Subgradients.thy [hunk 33]
```diff
@@ -0,0 +1,371 @@
...
+  imports "RLM_stable"
+begin
+
+
+definition subgradient :: "  ('a::euclidean_space \<Rightarrow> real) \<Rightarrow> 'a::euclidean_space set \<Rightarrow>
```
Comment (maxhaslbeck) [inline Subgradients.thy:6]: document these definitions

File: Subgradients.thy [hunk 34]
```diff
@@ -0,0 +1,371 @@
+theory Subgradients
+  imports "RLM_stable"
+begin
+
```
Comment (maxhaslbeck) [inline Subgradients.thy:4]: document and section the theory!

File: Subgradients.thy [hunk 35]
```diff
@@ -0,0 +1,371 @@
...
+      then have "\<rho> * e' \<ge> f ?u - f w" by auto
+      then  have " \<rho> * e' \<ge>  e' * norm v" using `f ?u - f w \<ge> e' * norm v`
+        by linarith
+      then show "\<rho> \<ge> norm v"
+        using \<open>0 < e'\<close> by auto
+    qed
+  qed
+qed
+
+lemma
```
Comment (maxhaslbeck) [inline Subgradients.thy:164]: lemma does not have a name, is it ever used?

File: Subgradients.thy [hunk 36]
```diff
@@ -0,0 +1,371 @@
...
+  assumes "\<rho> \<ge> 0"
+  assumes " open A"
+  assumes "convex A"
+  shows "convex_on A f \<longleftrightarrow> (\<forall> w \<in> A. \<exists> v \<in> A. \<forall> u \<in> A. f (u) \<ge> f(w) + (inner (u-w) v)) "
+  sorry
+
+
+
+
+lemma 147 :
```
Comment (maxhaslbeck) [inline Subgradients.thy:30]: can you find a better name for it, and add comment that it relates to lemma 14.7?

File: Subgradients.thy [hunk 37]
```diff
@@ -0,0 +1,371 @@
...
+      finally show "hinge_loss u (x,y)  \<ge> 
+      hinge_loss w (x,y) + (inner (u-w) (- y *\<^sub>R x))" by auto
+    qed
+    then show ?thesis
+      by blast
+  qed
+qed
+
+
+lemma abv:
```
Comment (maxhaslbeck) [inline Subgradients.thy]: non descriptive name

File: Subgradients.thy [hunk 38]
```diff
@@ -0,0 +1,371 @@
...
+      by (metis (mono_tags, hide_lams) \<open>max 0 ((1 - t) * (1 - snd z * (x \<bullet> fst z)) + t * (1 - snd z * (y \<bullet> fst z))) = max 0 (1 - t - (1 - t) * snd z * (x \<bullet> fst z) + t - t * snd z * (y \<bullet> fst z))\<close> \<open>max 0 ((1 - t) * (1 - snd z * (x \<bullet> fst z)) + t * (1 - snd z * (y \<bullet> fst z))) \<le> (1 - t) * max 0 (1 - snd z * (x \<bullet> fst z)) + t * max 0 (1 - snd z * (y \<bullet> fst z))\<close> \<open>max 0 (1 - (1 - t) * snd z * (x \<bullet> fst z) + - t * snd z * (y \<bullet> fst z)) = max 0 (1 - snd z * ((1 - t) *\<^sub>R x \<bullet> fst z) - snd z * (t *\<^sub>R y \<bullet> fst z))\<close> \<open>max 0 (1 - snd z * ((1 - t) *\<^sub>R x \<bullet> fst z) - snd z * (t *\<^sub>R y \<bullet> fst z)) = max 0 (1 - snd z * ((1 - t) *\<^sub>R x \<bullet> fst z + t *\<^sub>R y \<bullet> fst z))\<close> \<open>max 0 (1 - t - (1 - t) * snd z * (x \<bullet> fst z) + t - t * snd z * (y \<bullet> fst z)) = max 0 (1 - (1 - t) * snd z * (x \<bullet> fst z) + - t * snd z * (y \<bullet> fst z))\<close> add.commute  inner_commute inner_left_distrib inner_real_def inner_scaleR_left inner_scaleR_right)
```
Comment (maxhaslbeck) [inline Subgradients.thy]: shorten this!

File: Subgradients.thy [hunk 39]
```diff
@@ -0,0 +1,371 @@
...
+
+lemma hinge_loss_convex:
+  fixes H :: "'a::euclidean_space set"
+    and D :: "('a \<times> real) pmf"
+  shows "\<forall>z \<in> D. convex_on H (\<lambda> h. hinge_loss h z)"
+proof
+  fix z
+  assume "z \<in> D"
+  show " convex_on H (\<lambda> h. hinge_loss h z)" unfolding hinge_loss.simps
+  proof(rule)+
```
Comment (maxhaslbeck) [inline Subgradients.thy]: what about the following, and then using `a` and `b` instead of `fst z`and `snd z`
```suggestion
lemma hinge_loss_alt: "hinge_loss w (a,b) = (max 0 (1 - b * inner w a))"
  by simp

lemma hinge_loss_convex:
  fixes H :: "'a::euclidean_space set"
    and D :: "('a × real) pmf" 
  assumes "(a,b)∈D"
  shows "convex_on H (λ h. hinge_loss h (a,b))"
  unfolding hinge_loss_alt 
proof (rule)+
```

File: svm.thy [hunk 40]
```diff
@@ -0,0 +1,353 @@
+theory svm
```
Comment (maxhaslbeck) [inline svm.thy]: please adhere to [theory documentation guidelines](https://isabelle.systems/conventions/theory_documentation.html), at least add creator, summary and main definitions/theorems 

Spell out SVM, use Upper_Snake_Case

File: svm.thy [hunk 41]
```diff
@@ -0,0 +1,353 @@
...
+
+definition soft_svm_argmin :: " ('b \<times> (nat \<Rightarrow> real) )set" where
+  "soft_svm_argmin  = {h. is_arg_min (soft_svm) (\<lambda> x . fst x\<in>H \<and> (\<forall>i\<in>{0..<n}. snd (S i) * (inner (fst x) (fst (S i))) \<ge> 1 - (snd x i)) \<and> 
+    (\<forall> i\<in>{0..<n}. snd x i \<ge> 0) ) h}"
+
+definition soft_svm_mine :: "'b \<times> (nat \<Rightarrow> real)" where
+  "soft_svm_mine  = (SOME h. h \<in> soft_svm_argmin)"
+
+
+lemma asd :
```
Comment (maxhaslbeck) [inline svm.thy]: refactor name

File: svm.thy [hunk 42]
```diff
@@ -0,0 +1,353 @@
...
+    soft_svm (SOME h. is_arg_min (soft_svm) ?P h)" using  12  
+    by (smt is_arg_min_linorder someI_ex)
+  have "(SOME h. is_arg_min (soft_svm) ?P h) = soft_svm_mine" 
+    unfolding soft_svm_mine_def unfolding soft_svm_argmin_def
+    by simp
+  then show ?thesis using 13 by auto
+
+qed
+
+lemma
```
Comment (maxhaslbeck) [inline svm.thy]: this is corollary 15.7?

File: svm.thy [hunk 43]
```diff
@@ -0,0 +1,353 @@
...
+
+lemma asd1 :
+  assumes " h \<in> H"
+  assumes "is_arg_min (soft_svm) (\<lambda> x . fst x\<in>H \<and> 
+  (\<forall>i\<in>{0..<n}. snd (S i) * (inner (fst x) (fst (S i))) \<ge> 1 - (snd x i)) \<and> 
+    (\<forall> i\<in>{0..<n}. snd x i \<ge> 0) ) (h, ksi)"
+  shows "\<forall>i\<in>{0..<n}. ksi i = hinge_loss h (S i)" using asd 
+  using assms(1) assms(2) by auto
+
+lemma
```
Comment (maxhaslbeck) [inline svm.thy]: this is claim 15.5? the proof needs to be refactored, it is hard to follow! how does it relate to the informal proof in the book?

**Review by maxhaslbeck — CHANGES_REQUESTED** [review]
General Comments:
- you finished claim 15.5 and corollary 15.7 as you planned. Well done!
- You often use HOL quantifiers for lemmas, it is more general to use Pure quantifiers, as you do not need to use introduction rules when using the lemmas and can instantiate the variables more easily
- use long lemma format, to avoid writing the goal and assumptions twice
- some proofs need to be refactored, the theories need to be documented and section
- find many comments in the respective theories

File: RLM_learn.thy (hunk 18 above)
Comment (RaliDardjonova) [reply RLM_learn.thy:187]: Yes it is, not sure at the moment. I think I wasn't sure about some implementations and in the end, it turned out that it could be this easy. Oh wow, I am impressed :D
